### PR TITLE
Normalize images handled by Pillow in imread

### PR DESCRIPTION
**Pull request by creitiv**
It seems like PNG images loaded by `image.imread` are normalized to `[0, 1]` by default which behavior can be found in `src/_png.cpp:644`. This doesn't seems to apply for image types handled by Pillow instead of the builtin loaders. This fact lead to a lot of confusion and errors when I was working with different image types recently and the `imread` have unexpectedly yield different value ranges for different images.

Therefore it probably makes sense to normalize images, namely the one loaded by Pillow, to values from 0 to 1 to avoid further confusion, although this is not states in the documentation explicitly.

The referenced commit contains a simple fix for that as well as a test using two images of type `png` and `jpeg`.

### CODE REVIEW

File: lib/matplotlib/image.py [hunk 1]
```diff
@@ -1200,7 +1200,11 @@ def pilread(fname):
         except ImportError:
             return None
         with Image.open(fname) as image:
-            return pil_to_array(image)
+            array = pil_to_array(image)
```
Comment (tacaswell) [inline lib/matplotlib/image.py]: can PIL ever come back with negative numbers?

File: lib/matplotlib/image.py [hunk 2]
```diff
@@ -1200,7 +1200,11 @@ def pilread(fname):
...
+            # Since the handler for PNGs normalizes the image to values from 0 to 1
+            # this should be done for all other images too as it might lead to
+            # confusion if different value ranges appear for different image types.
+            return array / float(array.max())
```
Comment (tacaswell) [inline lib/matplotlib/image.py]: should this be `return (array - array.min()) / (array.max() - array.min())` ?

**Comment by tacaswell** [comment]
This is an API change, but given that the behavior is not documented, I am leaning towards it being in 2.1 if we decide to merge this.

More so than the range change, I am concerned about the possible change in the data type of the returned data.

`imread` is one of the parts of the API we maintain for historical reasons, but if we were starting from scratch would not include.

I am a soft :-1: on merging this.  I agree that the API that we have is less than great, but am concerned about the damage of changing it.  This fails an existing test which asserts that a 16bit tiff comes back as `uint16`.

I would suggest using something like http://scikit-image.org/docs/dev/api/skimage.io.html#imread (which has a bool flag to control exactly this) or https://pypi.python.org/pypi/imageio instead of using our `mpl.image.imread`

**Comment by tacaswell** [comment]
Sorry, forgot to add:  

Thank you for your work!

I hate writing negative reviews :/

**Comment by WeatherGod** [comment]
Yeah, I am wondering if clarifying the docs might be the better approach
here? Certainly, we could even direct users to other readers that may be
more better behaved? IIRC, the goal is to be able to essentially round-trip
stuff through imshow & imread. That might be a bit at odds with the goals
of a well-behaved imread tool from skimage and friends.

File: lib/matplotlib/image.py (hunk 2 above)
Comment (QuLogic) [inline lib/matplotlib/image.py]: Surely neither of these are correct; they will stretch the image contrast. It should be divided by the maximum possible value of the datatype.

**Comment by QuLogic** [comment]
The email used on your commit is invalid; you probably didn't intend that.

**Comment by creitiv** [comment]
> Yeah, I am wondering if clarifying the docs might be the better approach
here?

Also think the same. My confusion also came more from the fact that the behavior was not documented, than the behavior itself.

> Certainly, we could even direct users to other readers that may be
more better behaved?

I think this is a good idea. `imread` is a fast way to load single images when you just want to display them in a plot or do simply operations on them, but in this particular case I was using it to load large sets of images with various types for what there are obviously better methods and approaches. Such hints for the usage should really get included in the documentation.

> I hate writing negative reviews :/

Don't worry. I haven't contributed in that much projects on Github yet and it's always good to learn about what maintainers expect and how the workflow looks like.

This was also more or less a fast made heartless fix without really taking further consequences in mind as I currently don't have so much time to think about such things.

I think I will close this PL for now and maybe come up with a improved one in the next time.

Thank you for the discussion!

**Pull request closed without merging**